beacon-chain/sync/subscriber.go

```diff
@@ -198,8 +198,44 @@ func (s *Service) spawn(f func()) {
 	}
 }
 
+type subscriptionRegistrations struct {
+	sync.RWMutex
+	digests map[[4]byte]struct{}
+}
+
+// holdSubscriptionRegistration ensures that only one goroutine can register subscriptions.
+// It returns a boolean indicating whether the caller should proceed with registration,
+// and a callback function to release the lock when done.
+func (s *Service) holdSubscriptionRegistration(digest [4]byte) (bool, func()) {
+	// lazy init
+	if s.subRegs == nil {
+		s.subRegs = &subscriptionRegistrations{digests: make(map[[4]byte]struct{})}
```

Comment:
Will panic if `holdSubscriptionRegistration` is called concurrently with ` s.subRegs` unset.

Reply:
EDIT: I delete this comment.